### ISSUE TITLE
New lint suggestion: unused call to enumerate()

### ISSUE THREAD

**Issue by dbr** [rust-lang/rust-clippy#7527]
### What it does

Warn in cases where enumerate is called, but the enumerated value is ignored (`_i`)

### Categories (optional)

- Kind: I guess "complexity"

*What is the advantage of the recommended code over the original code*

Removes an unnecessary call, which introduces more code in what could otherwise be a simple for loop.

This can happen if `.enumerate()` is needed, then in the process of refactoring the index is no longer used - someone might silence the "unused variable warning" by putting an underscore in front of the first item in the tuple (something I think `cargo fix` will do automatically)

### Drawbacks

None.

### Example

```rust
let things = vec!["a", "b", "c"];
for (_i, item) in things.iter().enumerate() {
}
```

Could be written as:

```rust
let things = vec!["a", "b", "c"];
for item in things {
}
```

**Comment by r00ster91** [comment]
I'm not convinced we should have this lint. If the user explicitly added an underscore for the index then there is a reason. Maybe they want to change it later or some other reason. They wouldn't want a warning for that just like how you wouldn't want a warning for `let _unused = 1;`. If you argue that the lint is supposed to remind the user in case they forgot that this index variable is underscored/unused (and the `enumerate` can be removed), then the same argument could be used for `let _unused = 1;` which raises the question should Clippy emit a warning for every unused underscored variable? A warning like `hey I think you forgot about this variable. you probably want to use this`? No, I don't think we would want that and IMO that same thing applies here: there's a reason the user used an underscore. The compiler explicitly tells you to prefix your unused variables with an underscore and then it won't bother you anymore. We don't know why but it is intentional.

**Comment by matthiaskrgr** [comment]
If I go into my cargo cache at ` ~/.cargo/registry/src/github.com-1ecc6299db9ec823 ` and search the ~3000 crates I have extracted there for the pattern via `grep -r "enumerate"| grep "for (_\|, _.) in"` there are only around 54 matches.

Most of them seem to come from benchmarks or tests of the itertools crate.

**Comment by camsteffen** [comment]
> Maybe they want to change it later or some other reason.

This is not a good reason not to lint. Clippy's job is to lint code as it is written, not guess what it might be some day.

> then the same argument could be used for `let _unused = 1;` which raises the question should Clippy emit a warning for every unused underscored variable?

Actually I think a lint for `let _unused = 1;` may be a good idea - we can detect that the expression `1` has no side effects, and so that statement is useless. It is not the same as linting every unused variable. A valid use case for an unused variable is `let _ignore_result = do_something_useful();`. In the case of `enumerate`, we can detect that it can be removed without changing the semantics of the code, so this is a perfectly valid lint.

**Comment by dbr** [comment]
> Clippy's job is to lint code as it is written, not guess what it might be some day.

I think that is a good point

Additionally:
- the `_` might be there for a good reason, or it might be "because that warning told me to add an underscore", or added by `cargo fix` which I think was the case with when I spotted this in my own code base - both of these seem like cases where a linter message would be beneficial
- there is some existing lints which are triggered by `_unused` variable, main one which springs to mind is the [warning to use `for k in map.keys() {` instead of `for (k, _) in &map {`](https://rust-lang.github.io/rust-clippy/master/#for_kv_map), which seems quite conceptually similar to this proposed lint

I certainly don't think it's the most important lint, but seems like it'd be of use.

Further, I wonder if the compiler is smart enough to optimize away any overhead from `.enumerate()` in this case where the counter is unsed :thinking: 

Doesn't give me a good idea of scale compared to other linted patterns, but this search is quite interesting:
https://sourcegraph.com/search?q=context:global+lang:rust+for%5C+%5C%28_.*%3F%2C+.*%5C%29%5C+in%5C+.*.enumerate%28%29+%7B&patternType=regexp
It does come up quite a bit, most obviously in cases where there was probably some `dbg!(idx, item)` debug message at some point

**Comment by r00ster91** [comment]
> > Maybe they want to change it later or some other reason.
> 
> This is not a good reason not to lint. Clippy's job is to lint code as it is written, not guess what it might be some day.
> 
> > then the same argument could be used for `let _unused = 1;` which raises the question should Clippy emit a warning for every unused underscored variable?
> 
> Actually I think a lint for `let _unused = 1;` may be a good idea - we can detect that the expression `1` has no side effects, and so that statement is useless. It is not the same as linting every unused variable. A valid use case for an unused variable is `let _ignore_result = do_something_useful();`. In the case of `enumerate`, we can detect that it can be removed without changing the semantics of the code, so this is a perfectly valid lint.

Now I kind of agree. I thought underscored variables should always strictly be left alone and have no warnings whatsoever but in reality it certainly makes sense to apply lints to even underscored variables if they have no side effects. I made #7545.

**Comment by y21** [comment]
This lint exists today.

Output for the code in the example
```
warning: you seem to use `.enumerate()` and immediately discard the index
 --> src/main.rs:5:19
  |
5 | for (_i, item) in things.iter().enumerate() {
  |                   ^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unused_enumerate_index
  = note: `#[warn(clippy::unused_enumerate_index)]` on by default
help: remove the `.enumerate()` call
  |
5 | for item in things.iter() {
  |     ~~~~    ~~~~~~~~~~~~~
```
